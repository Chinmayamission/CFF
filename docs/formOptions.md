```
"formOptions": {
    "confirmationEmailInfo": ...,
    "showConfirmationPage": false,
    "successMessage": "<h1>Form submission success</h1><h2>Academician Form</h2><p>You will receive an email with confirmation shortly. Thank you!</p>"
  }
```


{
  "center": 1,
  "cff_permissions": {
    "cff:cognitoIdentityId:us-east-1:1e3aa7b7-b042-4834-98f1-7915985c39a5": {
      "owner": true
    }
  },
  "couponCodes": {},
  "date_created": "2018-05-20T03:59:32.211390",
  "date_last_modified": "2018-05-20T04:16:30.579101",
  "formOptions": {
    "confirmationEmailInfo": {
      "cc": null,
      "from": "webmaster@chinmayamission.com",
      "fromName": "CCMT Webmaster",
      "subject": "Academician Form - We have received your response",
      "template": {
        "html": "<img src=https://i.imgur.com/a9jf89X.png width=100%><h1>Confirmation</h1><h2>Academician Form</h2>Thank you for  submitting the form. This is a confirmation that we have received your response.<br><br><table>{% for key, value in response.items() %}<tr><th>{{key}}</th><td>{{value}}</td></tr>{% endfor %}</table>"
      },
      "toField": "email"
    },
    "showConfirmationPage": false,
    "successMessage": "<h1>Form submission success</h1><h2>Academician Form</h2><p>You will receive an email with confirmation shortly. Thank you!</p>"
  },
  "id": "e211731b-97f4-40ff-8ff6-9658d711d1a0",
  "name": "uischema and schema form -- Academicians Contact Form",
  "schema": {
    "definitions": {
      "name": {
        "properties": {
          "first": {
            "classNames": "twoColumn",
            "title": "First Name",
            "type": "string"
          },
          "last": {
            "classNames": "twoColumn",
            "title": "Last Name",
            "type": "string"
          }
        },
        "title": " ",
        "type": "object"
      }
    },
    "description": "Contact form.",
    "properties": {
      "area_expertise": {
        "title": "Area of Expertise",
        "type": "string"
      },
      "books_authored": {
        "format": "textarea",
        "title": "Books Authored",
        "type": "string"
      },
      "contact_name": {
        "properties": {
          "first": {
            "title": "First Name",
            "type": "string"
          },
          "last": {
            "title": "Last Name",
            "type": "string"
          }
        },
        "title": " ",
        "type": "object",
        "ui:order": [
          "first",
          "last"
        ]
      },
      "email": {
        "format": "email",
        "type": "string"
      },
      "papers_authored": {
        "format": "textarea",
        "title": "Papers Authored",
        "type": "string"
      },
      "phone": {
        "title": "Phone Number",
        "type": "string"
      }
    },
    "required": [
      "contact_name",
      "email",
      "area_expertise"
    ],
    "title": "Form with uiSchema and schema inline (Academicians Contact Form)",
    "type": "object"
  },
  "uiSchema": {
    "contact_name": {
      "first": {
        "classNames": "col-12 col-sm-6"
      },
      "last": {
        "classNames": "col-12 col-sm-6"
      },
      "ui:order": [
        "first",
        "last"
      ]
    },
    "phone": {
      "ui:placeholder": "Phone Number",
      "ui:widget": "phone"
    },
    "ui:order": [
      "contact_name",
      "email",
      "phone",
      "area_expertise",
      "books_authored",
      "papers_authored"
    ]
  },
  "version": 1
}