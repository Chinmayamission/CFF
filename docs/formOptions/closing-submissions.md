To close submissions, set the following values in formOptions to false: (both these attributes default to `true`):

```json
{
  "responseSubmissionEnabled": false,
  "responseModificationEnabled": false
}
```

The user will then, upon viewing the form, see a message like this:

![form closed](img/form closed.png)