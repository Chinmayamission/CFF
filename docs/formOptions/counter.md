You can give a numeric counter to each response as it gets added to the form. This is useful to create a "human-readable" numeric id that people can then see in their confirmation email and use to look responses up.

To do this, in `formOptions` add:

```json
{
    "counter": {
        "enabled": true
    }
}
```

Then, the `counter` field on the response will be incremented, starting with 1 for the first response, etc. This field is assigned directly on the response (as opposed to on the response value), so it is stored independently of the form data or associated schema, and it is not editable by users or admins.

!!! note
    The numbers are only assigned once upon response creation. If a response is deleted, then that number will be skipped.

## Adding a counter to the response table
To add the counter to the response table, enter `COUNTER` in the header value and it will be added.

## Adding a counter to a confirmation email
To add a counter to the confirmation email (or any jinja template), just use `{{counter}}` to display it.
