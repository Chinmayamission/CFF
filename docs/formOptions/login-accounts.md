By default, users do not need to login to fill out forms. To set up user login for a form,
set `formOptions.loginRequired` to be equal to true:

```json
{
  "loginRequired": true
}
```

## General flow

When `loginRequired` is set to true:

- When going to the form link, users will be prompted to log in if they are not logged in already.
- When submitting the form, the user's response will be associated with them.
- Each user can only submit one response per form. When the user goes back to the form, they will
be presented with their existing response, which they can then edit (if form editing is enabled).

## Predicates

When a form has `loginRequired` set to true, the predicates feature can also be enabled to import
data from previous iterations of a particular form. See [Predicates](predicates.md) for more information.