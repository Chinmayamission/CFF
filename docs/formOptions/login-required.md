To enable form login, add the following to formOptions:

```json
{
    "loginRequired": true
}
```

- Each form will be associated with the user who filled it out.
- *Only* that user may fill out that form.
- Modify links will send the user to a login screen if they have not already logged in, and when they log in, they will be shown their unique response for that form if present.

## Login Lookup
If you need to transition an existing anonymous-access form to a `loginRequired` form, set the `loginLookup` attribute.

```json
{
    "loginRequired": true,
    "loginLookup": true
}
```

Now, going to the form or a response page will show the login screen. However, the user will not be able to sign up for a new account, but will rather be prompted for their email address to look up an existing registration. Then, once they confirm their email, they will be redirected to the form.