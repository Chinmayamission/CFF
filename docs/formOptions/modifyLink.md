By default, a response has a field `modify_link` that is set to the original URL where the form is hosted, or if the form is hosted in an iframe, the parent's URL. However, especially if the form is hosted on an iframe in another website's domain, the modify_link may not properly give the entire path of the host website. To fix this, you can specify `formOptions.modifyLink`, which will set the `modify_link` of newly created responses to the configured link:

```json
{
    "modifyLink": "https://forms.ashwin.run/form/123
}
```
