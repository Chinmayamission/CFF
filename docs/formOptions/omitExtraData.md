The `formOptions.omitExtraData` property can be set to true. This sets the `omitExtraData` prop on the form, which causes extra form data values that are not in any form field will be removed upon submit. By default, this prop is set to false.

This is important when there are complex forms with dependencies, and you do not want extra data from unselected dependencies to show up. See the [react-jsonschema-form reference](https://react-jsonschema-form.readthedocs.io/en/latest/api-reference/form-props/#omitextradata) for more details.

```json
{
  "formOptions": {
    "omitExtraData": true
  }
}
```