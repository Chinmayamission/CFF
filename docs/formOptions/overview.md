FormOptions is a JSON object that can be used to configure other aspects of the form, such as:

- Payment integration
- Confirmation email template configuration
- Whether a form is closed or not