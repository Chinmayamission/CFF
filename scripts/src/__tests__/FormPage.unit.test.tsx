import React from 'react';
import store from "src/store";
import FormPage from 'src/form/FormPage';
import schema from './schema.json';
import uiSchema from './uiSchema.json';
import formOptions from './formOptions.json';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, render } from 'enzyme';

Enzyme.configure({ adapter: new Adapter() });

// https://hackernoon.com/testing-react-components-with-jest-and-enzyme-41d592c174f

let form_preloaded = {schema, uiSchema, formOptions};

it('renders FormPage correctly', () => {
  const wrapper = shallow(
    <FormPage store={store} formId={"test123"} form_preloaded={form_preloaded} />
  ).dive();
  expect(wrapper).toMatchSnapshot();
}); 