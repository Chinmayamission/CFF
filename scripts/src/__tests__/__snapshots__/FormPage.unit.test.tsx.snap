// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FormPage correctly 1`] = `
<div
  className="ccmt-cff-Page-FormPage "
>
  <CustomForm
    formData={Object {}}
    onChange={[Function]}
    onSubmit={[Function]}
    paymentCalcInfo={
      Object {
        "currency": "USD",
        "items": Array [
          Object {
            "amount": "150",
            "description": "With no children",
            "name": "2018-19 CMA Marietta Balavihar Family Registration",
            "quantity": "!donation and parents == 2 and children == 0",
          },
          Object {
            "amount": "300",
            "description": "With one child",
            "name": "2018-19 CMA Marietta Balavihar Family Registration",
            "quantity": "!donation and parents == 2 and children == 1",
          },
          Object {
            "amount": "450",
            "description": "With two or more children",
            "name": "2018-19 CMA Marietta Balavihar Family Registration",
            "quantity": "!donation and parents == 2 and children >= 2",
          },
          Object {
            "amount": "2000",
            "description": "Includes: CMA family membership for 1 year, including Bala Vihar fees; Complimentary tickets to fund-raising events benefiting CMA for 1 year",
            "name": "2018-19 CMA Marietta Balavihar Registration - One-time donation",
            "quantity": "donation",
          },
        ],
      }
    }
    schema={
      Object {
        "definitions": Object {
          "name": Object {
            "properties": Object {
              "first": Object {
                "title": "First Name",
                "type": "string",
              },
              "last": Object {
                "title": "Last Name",
                "type": "string",
              },
            },
            "required": Array [
              "first",
              "last",
            ],
            "title": " ",
            "type": "object",
          },
        },
        "description": "<strong>Note:</strong> This form is for registration at the Marietta location. If you need to register for the Alpharetta location, <a href='https://www.chinmayamissionalpharetta.org/registration/'>please go to the Alpharetta registration form</a>.<br><br>",
        "properties": Object {
          "accept_terms": Object {
            "description": "You will receive our weekly newsletter and teacher communication at the email address you have given. Please be aware of this.<br>You can edit your cell phone, email address, & street address.",
            "title": "Terms and Conditions",
            "type": "boolean",
          },
          "address": Object {
            "properties": Object {
              "city": Object {
                "classNames": "threeColumn",
                "type": "string",
              },
              "line1": Object {
                "title": "Address Line 1",
                "type": "string",
              },
              "line2": Object {
                "title": "Address Line 2",
                "type": "string",
              },
              "state": Object {
                "classNames": "threeColumn",
                "enum": Array [
                  "GA",
                ],
                "type": "string",
              },
              "zipcode": Object {
                "classNames": "threeColumn",
                "type": "string",
              },
            },
            "required": Array [
              "line1",
              "city",
              "state",
              "zipcode",
            ],
            "state": Object {
              "classNames": "col-12 col-sm-6 col-md-4",
            },
            "type": "object",
            "zipcode": Object {
              "classNames": "col-12 col-sm-6 col-md-4",
            },
          },
          "children": Object {
            "items": Object {
              "properties": Object {
                "allergies": Object {
                  "title": "Allergies (any we should know of)?",
                  "type": "string",
                },
                "dob": Object {
                  "format": "date",
                  "title": "Date of Birth",
                  "type": "string",
                },
                "email": Object {
                  "format": "email",
                  "title": "Email (if any)",
                  "type": "string",
                },
                "grade": Object {
                  "enum": Array [
                    "Pre K",
                    "KG",
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                    "6",
                    "7",
                    "8",
                    "9",
                    "10",
                    "11",
                    "12",
                  ],
                  "title": "Grade in September 2018",
                  "type": "string",
                },
                "interests": Object {
                  "items": Object {
                    "enum": Array [
                      "Art",
                      "Singing",
                      "Music",
                      "Dance",
                      "Sports",
                      "Cooking",
                      "Writing",
                      "Coding",
                    ],
                    "type": "string",
                  },
                  "title": "Tell us about your interests",
                  "type": "array",
                  "uniqueItems": true,
                },
                "language_class": Object {
                  "description": "Minimum requirement for Language Classes:<br>Sanskrit=4th grade and up<br>Hindi: Grade 1 and up<br>Gujarati=Toddler and up<br>All other languages=KG and up",
                  "enum": Array [
                    "Gujarati",
                    "Hindi",
                    "Sanskrit",
                    "Tamil",
                    "Telegu",
                    "Malayalam",
                    "Kannada",
                    "Marathi",
                  ],
                  "title": "Language Class",
                  "type": "string",
                },
                "name": Object {
                  "$ref": "#/definitions/name",
                },
                "previous": Object {
                  "description": "",
                  "title": "Have you previously attended Balavihar?",
                  "type": "boolean",
                },
              },
              "required": Array [
                "name",
                "grade",
                "dob",
              ],
              "title": "Child Information",
              "type": "object",
            },
            "minItems": 1,
            "type": "array",
          },
          "donation": Object {
            "description": "This includes:<ul><li>CMA family membership for 1 year, including Bala Vihar fees</li><li>Complimentary tickets to fund-raising events benefiting CMA for 1 year</li></ul>",
            "title": "Would you like to do a one time donation of $2,000 for this year instead?",
            "type": "boolean",
          },
          "parents": Object {
            "items": Object {
              "properties": Object {
                "email": Object {
                  "format": "email",
                  "type": "string",
                },
                "name": Object {
                  "$ref": "#/definitions/name",
                },
                "phone": Object {
                  "title": "Phone Number",
                  "type": "string",
                },
                "previous": Object {
                  "title": "Have you previously been involved in Chinmaya Mission?",
                  "type": "boolean",
                },
                "volunteer": Object {
                  "enum": Array [
                    "Teaching",
                    "Backup teacher",
                    "Art activity",
                    "Music",
                    "Dance",
                    "Technology",
                    "Cooking",
                    "Event organization",
                  ],
                  "title": "Would you like to volunteer at CMA Marietta? If so, how?",
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "email",
              ],
              "title": " ",
              "type": "object",
            },
            "maxItems": 2,
            "minItems": 2,
            "type": "array",
          },
          "subscribe": Object {
            "description": "I would like to be informed by email about future events from Chinmaya Mission.",
            "title": "I would like to be informed about future Chinmaya Mission events by email.",
            "type": "boolean",
          },
        },
        "required": Array [
          "parents",
          "subscribe",
          "donation",
          "accept_terms",
        ],
        "title": "<img src='https://www.chinmayamissionalpharetta.org/wp-content/themes/cma/img/headerImg.png' style='width:50%;display:block;margin:auto;'><br>2018-19 Marietta Balavihar Registration Form",
        "type": "object",
      }
    }
    showPaymentTable={true}
    uiSchema={
      Object {
        "address": Object {
          "city": Object {
            "classNames": "col-12 col-sm-6 col-md-4",
          },
          "classNames": "col-12",
          "line1": Object {
            "classNames": "col-12 col-sm-6",
          },
          "line2": Object {
            "classNames": "col-12 col-sm-6",
          },
          "state": Object {
            "classNames": "col-12 col-sm-6 col-md-4",
          },
          "ui:order": Array [
            "line1",
            "line2",
            "city",
            "state",
            "zipcode",
          ],
          "zipcode": Object {
            "classNames": "col-12 col-sm-6 col-md-4",
          },
        },
        "children": Object {
          "classNames": "col-12",
          "items": Object {
            "allergies": Object {
              "classNames": "col-12 col-sm-6",
              "ui:widget": "textarea",
            },
            "dob": Object {
              "classNames": "col-12 col-sm-6",
            },
            "email": Object {
              "classNames": "col-12 col-sm-6",
            },
            "grade": Object {
              "classNames": "col-12 col-sm-6",
            },
            "interests": Object {
              "classNames": "col-12 col-sm-6",
              "ui:widget": "checkboxes",
            },
            "language_class": Object {
              "classNames": "col-12 col-sm-6",
            },
            "name": Object {
              "classNames": "col-12",
              "first": Object {
                "classNames": "col-12 col-sm-6",
              },
              "last": Object {
                "classNames": "col-12 col-sm-6",
              },
              "ui:order": Array [
                "first",
                "last",
              ],
            },
            "previous": Object {
              "classNames": "col-12",
              "ui:widget": "radio",
            },
            "required": Array [
              "name",
              "dob",
              "grade",
            ],
            "ui:order": Array [
              "name",
              "dob",
              "grade",
              "email",
              "language_class",
              "previous",
              "allergies",
              "interests",
            ],
          },
        },
        "contact_name": Object {
          "first": Object {
            "classNames": "col-12 col-sm-6",
          },
          "last": Object {
            "classNames": "col-12 col-sm-6",
          },
          "ui:order": Array [
            "first",
            "last",
          ],
        },
        "donation": Object {
          "classNames": "col-12",
          "ui:widget": "radio",
        },
        "geeta_class": Object {
          "classNames": "col-12",
          "ui:widget": "radio",
        },
        "parents": Object {
          "classNames": "col-12",
          "items": Object {
            "email": Object {
              "classNames": "col-12 col-sm-6 col-md-3",
            },
            "name": Object {
              "classNames": "col-12 col-md-6",
              "first": Object {
                "classNames": "col-12 col-sm-6",
              },
              "last": Object {
                "classNames": "col-12 col-sm-6",
              },
              "ui:order": Array [
                "first",
                "last",
              ],
            },
            "phone": Object {
              "classNames": "col-12 col-sm-6 col-md-3",
            },
            "previous": Object {
              "classNames": "col-12",
              "ui:widget": "radio",
            },
            "ui:order": Array [
              "name",
              "phone",
              "email",
              "previous",
              "volunteer",
            ],
          },
        },
        "subscribe": Object {
          "classNames": "col-12",
          "ui:widget": "radio",
        },
        "ui:cff:background": "#ff7600",
        "ui:order": Array [
          "parents",
          "children",
          "address",
          "subscribe",
          "donation",
          "accept_terms",
        ],
      }
    }
  />
</div>
`;
